New Refresh Token Rotation content (#8821)
auth0/docs: articles/tokens/concepts/token-storage.md

When the SPA calls multiple APIs that reside in a different domain, access and optionally refresh tokens are needed.

-  If the SPA backend can handle the API calls, handle tokens server-side using:
    - [Authorization Code Flow](/flows/concepts/auth-code)
    - [Authorization Code Flow with Proof Key for Code Exchange (PKCE)](/flows/concepts/auth-code-pkce)
    - [Hybrid Flow](/api-auth/grant/hybrid)

- If the SPA backend cannot handle the API calls, the tokens should be stored in the SPA backend but the SPA needs to fetch the tokens from the backend to perform requests to the API. A protocol needs to be established between the backend and the SPA to allow the secure transfer of the token from the backend to the SPA.

- If you have a SPA with **no** corresponding backend server, your SPA should request new tokens on login and store them in memory without any persistence. To make API calls, your SPA would then use the in-memory copy of the token.

### Browser in-memory scenarios

> Auth0 recommends that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) when storing tokens in memory to emulate private methods. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js). 

Reviewer: ```suggestion
Auth0 recommends storing tokens in browser memory as the most secure option. Using [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) to handle the transmission and storage of tokens is the best way to protect the tokens, as Web Workers run in a separate global scope than the rest of the application. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js) whose default storage option is in-memory storage leveraging Web Workers.

If you cannot use Web Workers, Auth0 recommends as an alternative that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) to emulate private methods.
```

- If you have a SPA with **no** corresponding backend server, your SPA should request new tokens on login and store them in memory without any persistence. To make API calls, your SPA would then use the in-memory copy of the token.

### Browser in-memory scenarios

Auth0 recommends storing tokens in browser memory as the most secure option. Using [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) to handle the transmission and storage of tokens is the best way to protect the tokens, as Web Workers run in a separate global scope than the rest of the application. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js) whose default storage option is in-memory storage leveraging Web Workers.

If you cannot use Web Workers, Auth0 recommends as an alternative that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) to emulate private methods.

::: warning
The in-memory method for browser storage **does not** provide persistence across page refreshes and browser tabs. 
:::

### Browser local storage scenarios

> Storing tokens in browser local storage provides persistence across page refreshes and browser tabs, however if an attacker can achieve running JavaScript in the SPA, they can retrieve the tokens stored in local storage using a cross-site scripting (XSS) attack. A vulnerability leading to a successful XSS attack can be either in the SPA source code or in any third-party JavaScript code (such as bootstrap, jQuery, or Google Analytics) included in the SPA.

Reviewer: ```suggestion
Storing tokens in browser local storage provides persistence across page refreshes and browser tabs, however if an attacker can achieve running JavaScript in the SPA using a cross-site scripting (XSS) attack, they can retrieve the tokens stored in local storage. A vulnerability leading to a successful XSS attack can be either in the SPA source code or in any third-party JavaScript code (such as bootstrap, jQuery, or Google Analytics) included in the SPA.
```

Author: I added this manually - as I had done it already based on your Slack comment.

When the SPA calls only an API that is served from a domain that can share cookies with the domain of the SPA, no tokens are needed. OAuth adds additional attack vectors without providing any additional value and should be avoided in favor of a traditional cookie-based approach. For an overview of this approach and an example implementation, see [Single-Page App Authentication Using Cookies](/login/spa/authenticate-with-cookies).

When the SPA calls multiple APIs that reside in a different domain, access and optionally refresh tokens are needed.

-  If the SPA backend can handle the API calls, handle tokens server-side using:
    - [Authorization Code Flow](/flows/concepts/auth-code)
    - [Authorization Code Flow with Proof Key for Code Exchange (PKCE)](/flows/concepts/auth-code-pkce)
    - [Hybrid Flow](/api-auth/grant/hybrid)

- If the SPA backend cannot handle the API calls, the tokens should be stored in the SPA backend but the SPA needs to fetch the tokens from the backend to perform requests to the API. A protocol needs to be established between the backend and the SPA to allow the secure transfer of the token from the backend to the SPA.

> - If you have a SPA with **no** corresponding backend server, your SPA should request new tokens on login and store them in memory without any persistence. To make API calls, your SPA would then use the in-memory copy of the token.

Reviewer 2: > If you have a SPA

Throughout the docs there's a mix of "a SPA" and "an SPA" - I recommend having a sweep through and picking one style or the other, for consistency.

Author: Good catch. I'll do that.

### Browser in-memory scenarios

> Auth0 recommends storing tokens in browser memory as the most secure option. Using [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) to handle the transmission and storage of tokens is the best way to protect the tokens, as Web Workers run in a separate global scope than the rest of the application. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js) whose default storage option is in-memory storage leveraging Web Workers.

Reviewer: ```suggestion
Auth0 recommends storing tokens in browser memory as the most secure option. Using [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) to handle the transmission and storage of tokens is the best way to protect the tokens, as Web Workers run in a separate global scope than the rest of the application. 
```

Author: Comitted

> If you cannot use Web Workers, Auth0 recommends as an alternative that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) to emulate private methods.

Reviewer: ```suggestion
If you cannot use Web Workers, Auth0 recommends as an alternative that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) to emulate private methods.

 Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js) whose default storage option is in-memory storage. It leverages both Web Workers and JavaScript closures depending on the type of token.
```

Reviewer 2: > whose default storage option is in-memory storage leveraging Web Workers.

Just a question - the SPA SDK will leverage web workers where possible, except for IE11 where they are supported, but have other complications. Do we want to make reference to anything about that here, to the effect of "It leverages both Web Workers (in modern browsers), and..."

We could link to some readme docs that explains web workers and browser support specific to the SDK, in more detail.

Reviewer: It's not only IE11. Access Tokens can't be stored in Web Workers. This document is about tokens in general.
Linking to some readme docs is a good idea!

Reviewer 2: The document is, but this paragraph in particular is in the context of the SPA SDK. Perhaps it's not the right place for that kind of detail, but let's link to specifics.

Reviewer: Right and in the SPA SDK we use Web Workers for RT in modern browsers and JavaScript closures for ATs and for RTs in IE11. Am I missing something?

Reviewer 2: No you're correct, but the paragraph doesn't mention modern browsers - my question (and it was only a question) was whether or not it should.

- [Saving and Renewing Tokens for Android](/libraries/auth0-android/save-and-refresh-tokens)
- [Saving and Renewing Tokens for Swift](/libraries/auth0-swift/save-and-refresh-jwt-tokens)
- [Native/Mobile Apps Quickstarts](/quickstart/native)

### SPA backend scenarios

When the SPA calls only an API that is served from a domain that can share cookies with the domain of the SPA, no tokens are needed. OAuth adds additional attack vectors without providing any additional value and should be avoided in favour of a traditional cookie based approach.

When the SPA calls multiple APIs living in a different domain, access and optionally refresh tokens are needed.

-  If the SPA backend can handle the API calls, the tokens should be handled and stored similarly to a Web App scenario.

- If the SPA backend cannot handle the API calls, the tokens should be stored in the SPA backend but the SPA needs to fetch the tokens from the backend to perform requests to the API. A protocol needs to be established between the backend and the SPA to allow the secure transfer of the token from the backend to the SPA.

### Browser in-memory scenarios

> Auth0 recommends that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) when storing tokens in memory to emulate private methods. Use [auth0-sp-js](https://github.com/auth0/auth0-spa-js). This is the most secure method for browser storage, however it **does not** provide persistence across page refreshes and browser tabs. 

Reviewer: ```suggestion
Auth0 recommends that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) when storing tokens in memory to emulate private methods. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js). This is the most secure method for browser storage, however it **does not** provide persistence across page refreshes and browser tabs. 
```

### Browser local storage scenarios

Storing tokens in browser local storage provides persistence across page refreshes and browser tabs, however if an attacker can achieve running JavaScript in the SPA, they can retrieve the tokens stored in local storage using a cross-site scripting (XSS) attack. A vulnerability leading to a successful XSS attack can be either in the SPA source code or in any third-party JavaScript code (such as bootstrap, jQuery, or Google Analytics) included in the SPA.

Using browser local storage can be viable alternative to mechanisms that require retrieving the Access Token from an iframe and to cookie-based authentication across domains when these are not possible due to browser restrictions (e.g. ITP2).

::: panel Reduce Security Risks with Local Storage
If the SPA is using Implicit (although it is recommended that authorization code with PKCE is used) or Hybrid Flows, the absolute token expiration time can be reduced. This will reduce the impact of a reflected XSS (but not of a persistent one). To do so, go to **Dashboard > APIs > Settings > Token Expiration For Browser Flows (Seconds)**.

> Reduce the amount of third party JavaScript code included from a source outside your domain to the minimum needed (such as links to jQuery, Bootstrap, Google Analytics etc.) Reducing third-party JS code reduces the possibility of an XSS vulnerability. Performing Subresource Integrity checking (SRI) in third-party scripts (where possible) to verify that the resources fetched are delivered without unexpected manipulation is also more secure.

Reviewer: Let's add a link in Subresource Integrity checking (SRI): https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity

Author: Added
